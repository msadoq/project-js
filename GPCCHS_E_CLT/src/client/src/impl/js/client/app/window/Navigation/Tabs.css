.title {
  padding-right: 15px;
}

.close-button {
  position: absolute;
  right: 8px;
  top: 11px;
}
