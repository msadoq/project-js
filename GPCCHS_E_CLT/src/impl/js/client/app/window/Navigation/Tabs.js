import React, { Component, PropTypes } from 'react';
import { Nav, NavItem, Glyphicon } from 'react-bootstrap';
import styles from './Tabs.css';

export default class Tabs extends Component {
  static propTypes = {
    pages: PropTypes.array.isRequired,
    focusedPage: PropTypes.string,
    focusPage: PropTypes.func,
    addAndMount: PropTypes.func,
    removeAndUnmountPage: PropTypes.func,
  };
  constructor(...args) {
    super(...args);
    this.handleSelect = this.handleSelect.bind(this);
    this.handleClose = this.handleClose.bind(this);
  }
  handleSelect(eventKey) {
    if (eventKey === 'new') {
      return this.props.addAndMount();
    }

    this.props.focusPage(eventKey);
  }
  handleClose(e, pageId) {
    e.preventDefault();
    this.props.removeAndUnmountPage(pageId);
  }
  render() {
    const { pages, focusedPage } = this.props;

    return (
      <Nav bsStyle="tabs" activeKey={focusedPage} onSelect={this.handleSelect}>
        {pages.map(page =>
          <NavItem
            key={page.pageId}
            eventKey={page.pageId}
          >
            <div className={styles.title}>
              {page.title}
            </div>
            <div onClick={e => this.handleClose(e, page.pageId)} className={styles['close-button']}>
              <Glyphicon glyph="remove-circle" />
            </div>
          </NavItem>
        )}
        <NavItem eventKey="new">New page +</NavItem>
      </Nav>
    );
  }
}
