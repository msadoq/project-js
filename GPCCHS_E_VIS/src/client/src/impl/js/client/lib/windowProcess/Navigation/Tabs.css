.title {
  color: #708090;
  font-weight: bold;

}

.close {
  padding: 5px;
}
