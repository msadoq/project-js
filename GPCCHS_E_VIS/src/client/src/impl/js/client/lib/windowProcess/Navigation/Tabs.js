import React, { Component, PropTypes } from 'react';
import { Nav, NavItem, Button, Glyphicon } from 'react-bootstrap';
import diff from 'deep-diff';
import styles from './Tabs.css';
import debug from '../../../lib/common/debug/windowDebug';

const logger = debug('Tabs');

export default class Tabs extends Component {
  static propTypes = {
    pages: PropTypes.array.isRequired,
    focusedPageId: PropTypes.string,
    focusPage: PropTypes.func,
    addAndMount: PropTypes.func,
    removeAndUnmountPage: PropTypes.func,
  };

  componentWillReceiveProps(nextProps) {
    console.log('componentWillReceiveProps', diff(this.props, nextProps));
  }

  handleSelect = (eventKey) => {
    if (eventKey === 'new') {
      return this.props.addAndMount();
    }

    this.props.focusPage(eventKey);
  }

  handleClose = (e, pageId) => {
    e.preventDefault();
    e.stopPropagation();
    this.props.removeAndUnmountPage(pageId);
    e.stopPropagation();
  }

  render() {
    logger.debug('render');
    const { pages, focusedPageId } = this.props;
    return (
      <Nav bsStyle="tabs" activeKey={focusedPageId} onSelect={this.handleSelect}>
        {pages.map(page =>
          <NavItem
            key={page.pageId}
            eventKey={page.pageId}
          >
            <div className={styles.title}>
              {page.title}
              <Button
                bsStyle="link"
                onClick={e => this.handleClose(e, page.pageId)}
                className={styles.close}
              >
                <Glyphicon glyph="remove-circle" />
              </Button>
            </div>
          </NavItem>
        )}
        <NavItem eventKey="new">New page +</NavItem>
      </Nav>
    );
  }
}
