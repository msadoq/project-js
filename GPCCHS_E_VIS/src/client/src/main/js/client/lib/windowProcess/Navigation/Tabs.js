import React, { PureComponent, PropTypes } from 'react';
import { basename } from 'path';
import { Nav, NavItem, Button, Glyphicon, OverlayTrigger, Table, Popover } from 'react-bootstrap';
import { get } from 'common/parameters';
import DummyDrag from './DummyDrag';
import styles from './Tabs.css';

const popoverDraggingStyle = { display: 'none' };

function popoverHoverFocus(page) {
  return (
    <Popover id={page.pageId} title="Document properties">
      <Table>
        <tbody>
          {page.properties.length ?
            page.properties.map(
              (prop, idx) => <tr key={page.pageId.concat(idx)}>
                <td>{prop.name.value}</td>
                <td>{prop.value.value}</td>
              </tr>)
            : <tr><td>No FMD data</td></tr>}
          {page.oId && <tr><td>OID</td><td>{page.oId}</td></tr>}
          {page.absolutePath &&
            <tr><td>File name</td><td>{basename(page.absolutePath)}</td></tr>}
          {!page.absolutePath && page.path &&
            <tr><td>File name</td><td>{basename(page.path)}</td></tr>}
          {!page.absolutePath && !page.path && <tr><td>Unsaved file</td></tr>}
        </tbody>
      </Table>
    </Popover>
  );
}
function popoverHoverDragging(page) {
  return (
    <Popover id={page.pageId} style={popoverDraggingStyle} />
  );
}

export default class Tabs extends PureComponent {
  static propTypes = {
    pages: PropTypes.arrayOf(PropTypes.object).isRequired,
    focusedPageId: PropTypes.string,
    focusPage: PropTypes.func.isRequired,
    closePage: PropTypes.func.isRequired,
    moveTabOrder: PropTypes.func.isRequired,
  };

  handleSelect = (eventKey) => {
    if (eventKey) {
      this.props.focusPage(eventKey);
    }
  }

  handleClose = (e, pageId) => {
    e.preventDefault();
    e.stopPropagation();
    this.props.closePage(pageId);
    e.stopPropagation();
  }

  handleDragStart = (ev, pageId, key) => {
    // console.log('start');
    this.setState({ dragging: pageId });
    ev.dataTransfer.setData('pageId', pageId);
    ev.dataTransfer.setData('position', key);
  }

  handleDragStop = () => {
    this.setState({ dragging: null });
  //  console.log(e.clientX);
  }

  handleDragOver = (e) => {
    e.preventDefault();
    const ev = e.currentTarget.parentNode;
    if (ev.tagName.toLowerCase() === 'li') {
      ev.style.borderLeft = 'none';
      ev.style.borderRight = 'none';
      if (
        e.clientX - ev.getBoundingClientRect().left >
        ev.getBoundingClientRect().width / 2
      ) {
        ev.style.borderRight = '2px solid rgb(51,122,183)';
      } else {
        ev.style.borderLeft = '2px solid rgb(51,122,183)';
      }
    }
  }

  handleDragLeave = (e) => {
    const ev = e.currentTarget.parentNode;
    ev.style.borderLeft = 'none';
    ev.style.borderRight = 'none';
  }

  handleDrop = (e, pageId, key) => {
    const ev = e.currentTarget.parentNode;
    ev.style.borderLeft = 'none';
    ev.style.borderRight = 'none';
    let newKey = key;
    if (
      e.clientX - ev.getBoundingClientRect().left >
      ev.getBoundingClientRect().width / 2
    ) {
      newKey = key + 1;
    }
    this.props.moveTabOrder(parseInt(e.dataTransfer.getData('position'), 10), newKey);
  }

  render() {
    const { pages, focusedPageId } = this.props;
    let dumpColor = '';
    if (get('DUMP') === 'on') {
      dumpColor = styles.band;
    }

    return (
      <Nav bsStyle="tabs" activeKey={focusedPageId} onSelect={this.handleSelect} className={dumpColor}>
        {
          pages.map((page, key) =>
            <NavItem
              key={page.pageId}
              eventKey={page.pageId}
              onDragOver={this.handleDragOver}
              onDragLeave={this.handleDragLeave}
              onDrop={(e) => { this.handleDrop(e, page.title, key); }}
            >
              <div
                draggable
                onDragStart={(e) => { this.handleDragStart(e, page.title, key); }}
                onDragEnd={this.handleDragStop}
                onDrag={this.handleDrag}
              >
                <OverlayTrigger
                  className={styles.title}
                  overlay={
                    this.state &&
                    this.state.dragging ?
                      popoverHoverDragging(page, this.state.dragging) :
                      popoverHoverFocus(page)
                  }
                >
                  <div>
                    <span>{page.isModified ? page.title.concat(' *') : page.title}</span>
                    <Button
                      bsStyle="link"
                      onClick={e => this.handleClose(e, page.pageId)}
                      className={styles.button}
                    >
                      <Glyphicon
                        glyph="remove-circle"
                        className="text-danger"
                      />
                    </Button>
                  </div>
                </OverlayTrigger>
              </div>
            </NavItem>
          )
        }
        <DummyDrag />
      </Nav>
    );
  }
}
