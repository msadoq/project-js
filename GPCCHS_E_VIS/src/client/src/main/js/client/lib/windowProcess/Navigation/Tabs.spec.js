import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import Tabs from 'windowProcess/Navigation/Tabs';
import { shallow } from 'enzyme';

const propsStub = {
  pages: [
    {},
    {},
  ],
  windowId: '5c467cc7-bfb4-4fae-85a8-b0fe7e87dfde',
  detachWindow: true,
  attachWindow: '5c467cc7-bfb4-4fae-85a8-b0fe7e87dfde',
  focusedPageId: '5c467cc7-cezl-5dcs-f54s-kjdshnlskjd45',
  focusPage: () => {},
  askClosePage: () => {},
  movePageToWindow: () => {},
  moveTabOrder: () => {},
  pageDragEvent: () => {},
};

describe('windowProcess', () => {
  describe('windowProcess :: Navigation', () => {
    describe('windowProcess :: Navigation :: Tabs', () => {
      test('Tabs :: render', () => {
        const store = createStore(state => state, {});
        const tree = shallow(
          <Provider store={store}>
            <Tabs {...propsStub} />
          </Provider>
        );
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
