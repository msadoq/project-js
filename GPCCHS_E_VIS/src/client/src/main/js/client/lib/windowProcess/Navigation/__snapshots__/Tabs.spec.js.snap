// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`windowProcess windowProcess :: Navigation windowProcess :: Navigation :: Tabs Tabs :: render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(Symbol.observable): [Function],
          }
    }
>
    <Tabs
        askClosePage={[Function]}
        attachWindow="5c467cc7-bfb4-4fae-85a8-b0fe7e87dfde"
        detachWindow={true}
        focusPage={[Function]}
        focusedPageId="5c467cc7-cezl-5dcs-f54s-kjdshnlskjd45"
        movePageToWindow={[Function]}
        moveTabOrder={[Function]}
        pageDragEvent={[Function]}
        pages={
            Array [
                Object {},
                Object {},
              ]
        }
        windowId="5c467cc7-bfb4-4fae-85a8-b0fe7e87dfde"
    />
</Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "askClosePage": [Function],
      "attachWindow": "5c467cc7-bfb4-4fae-85a8-b0fe7e87dfde",
      "detachWindow": true,
      "focusPage": [Function],
      "focusedPageId": "5c467cc7-cezl-5dcs-f54s-kjdshnlskjd45",
      "movePageToWindow": [Function],
      "moveTabOrder": [Function],
      "pageDragEvent": [Function],
      "pages": Array [
        Object {},
        Object {},
      ],
      "windowId": "5c467cc7-bfb4-4fae-85a8-b0fe7e87dfde",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "askClosePage": [Function],
        "attachWindow": "5c467cc7-bfb4-4fae-85a8-b0fe7e87dfde",
        "detachWindow": true,
        "focusPage": [Function],
        "focusedPageId": "5c467cc7-cezl-5dcs-f54s-kjdshnlskjd45",
        "movePageToWindow": [Function],
        "moveTabOrder": [Function],
        "pageDragEvent": [Function],
        "pages": Array [
          Object {},
          Object {},
        ],
        "windowId": "5c467cc7-bfb4-4fae-85a8-b0fe7e87dfde",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenFourAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
